### PR TITLE
Add Linux door-lock app

### PR DESCRIPTION
**Pull request by dustin-crossman**
#### Problem
What is being fixed?  Examples:
* Adds bare-bones linux door-lock app

#### Change overview
* Linux door-lock app pulls in recently updated door-lock cluster
* Implement LockDoor and UnlockDoor in door lock cluster code
* Adds Linux door-lock app to github workflows

#### Testing
* Manual testing using chip-tool and all-cluster-app
* CI automated testing

### CODE REVIEW

File: .github/workflows/examples-linux-standalone.yaml [hunk 1]
```diff
@@ -128,6 +128,14 @@ jobs:
                     out/ota_requestor_debug/chip-ota-requestor-app \
                     /tmp/bloat_reports/
 
+            - name: Build example Standalone Door Lock App
+              timeout-minutes: 5
+              run: |
+                  scripts/examples/gn_build_example.sh examples/lock-app/linux out/lock_app_debug
```
Comment (andy31415) [inline .github/workflows/examples-linux-standalone.yaml:134]: could you add it (and use) build_examples.py instead?

File: examples/lock-app/linux/include/LockManager.h [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+class LockManager
+{
+public:
+    void Init();
+
+    bool Lock(const char* pin);
+    bool Unlock(const char* pin);
+
+private:
+    friend LockManager & LockMgr(void);
```
Comment (andy31415) [inline examples/lock-app/linux/include/LockManager.h:35]: why do we do this instead of providing something like a `static LockManager& Instance()` ?

**Comment by andy31415** [comment]
@dustin-crossman please also fix restyle issues (e.g. by running `./scripts/helpers/restyle-diff.sh`), I believe reviewers are only added after that check passes.

File: examples/lock-app/linux/main.cpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+using namespace chip;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::DoorLock;
+
+// App handles physical aspects of locking but not locking logic. That is it
+// should wait for door to be locked on lock command and return success) but
+// door lock server should check pin before even calling the lock-door
+// callback.
+bool emberAfPluginDoorLockOnDoorLockCommand(chip::EndpointId endpointId, const char * PINCode)
```
Comment (bzbarsky-apple) [inline examples/lock-app/linux/main.cpp:34]: This signature doesn't make sense to me.  PIN is an octet string, not a character string, in door lock cluster...  Same for unlock.

File: examples/lock-app/linux/main.cpp [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+void MatterPostAttributeChangeCallback(const chip::app::ConcreteAttributePath & attributePath, uint8_t mask, uint8_t type, uint16_t size, uint8_t * value)
+{
+    // TODO: Watch for LockState, DoorState, Mode, etc changes and trigger appropriate action
+    if (attributePath.mClusterId == DoorLock::Id)
+    {
+        emberAfDoorLockClusterPrintln("Door Lock attribute changed");
+    }
+}
+
+void emberAfDoorLockClusterInitCallback(EndpointId endpoint)
```
Comment (bzbarsky-apple) [inline examples/lock-app/linux/main.cpp:53]: This should be in the cluster impl, I'd think...  And call into some sort of app-provided API for app-specific init if needed.

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 5]
```diff
@@ -70,25 +73,25 @@ bool DoorLockServer::SetLockState(chip::EndpointId endpointId, DlLockState newLo
...
+bool DoorLockServer::SetActuatorEnabled(chip::EndpointId endpointId, bool newActuatorState)
 {
     auto actuatorState = static_cast<uint8_t>(newActuatorState);
 
-    emberAfDoorLockClusterPrintln("Setting Actuator State to '%hhu'", actuatorState);
+    emberAfDoorLockClusterPrintln("Setting Actuator Enabled State to '%hhu'", actuatorState);
 
-    bool status = Attributes::LockState::Set(endpointId, actuatorState);
-    if (!status)
+    bool status = Attributes::ActuatorEnabled::Set(endpointId, actuatorState);
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:91]: Pre-existing, but should not be bool.

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 6]
```diff
@@ -101,7 +104,7 @@ bool DoorLockServer::SetDoorState(chip::EndpointId endpointId, DlLockState newDo
     emberAfDoorLockClusterPrintln("Setting Door State to '%hhu'", doorState);
     bool status = Attributes::DoorState::Set(endpointId, doorState);
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:105]: Again, should not be bool.

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 7]
```diff
@@ -140,30 +143,115 @@ bool emberAfDoorLockClusterLockDoorCallback(chip::app::CommandHandler * commandO
...
-    emberAfSendImmediateDefaultResponse(EMBER_ZCL_STATUS_SUCCESS);
-    return true;
+    bool require_pin = false;
+    Attributes::RequirePINforRemoteOperation::Get(endpoint, &require_pin);
+
+    if(require_pin)
+    {
+        if(!commandData.pinCode.HasValue())
+        {
+            success = false; // Just to be explicit. success == false at this point anywyay
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:158]:
```suggestion
            success = false; // Just to be explicit. success == false at this point anyway
```

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 8]
```diff
@@ -140,30 +143,115 @@ bool emberAfDoorLockClusterLockDoorCallback(chip::app::CommandHandler * commandO
...
+        }
+        else
+        {
+            const char *cmd_pin = reinterpret_cast<const char*>(commandData.pinCode.Value().data());
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:162]: No, you can't do that.  The value is not null-terminated, and is not a char string anyway.  The might be no null anywhere around, or it might contain multiple nulls.

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 9]
```diff
@@ -140,30 +143,115 @@ bool emberAfDoorLockClusterLockDoorCallback(chip::app::CommandHandler * commandO
...
+            char *mPin = DoorLockServer::Instance().mPin;
+            if(strncmp(mPin, cmd_pin, 4) == 0)
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:164]: So what this _actually_ needs to do is go through the list of stored PINs.  Please add a TODO comment to that effect.

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 10]
```diff
@@ -140,30 +143,115 @@ bool emberAfDoorLockClusterLockDoorCallback(chip::app::CommandHandler * commandO
...
+            {
+                success = false;
+            }
+        }
+    }
+    else
+    {
+        const char *cmd_pin = commandData.pinCode.HasValue() ?
+            reinterpret_cast<const char*>(commandData.pinCode.Value().data()) :
+            NULL;
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:178]: This does not look right.  Not requiring PIN just means that if it's _not_ provided that's OK and the command executes.  If it's provided, it still needs to be checked.  So the logic in this function should probably test `pinCode.HasValue()` at toplevel, not `require_pin`, and only worry about `require_pin` if there is no PIN.

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 11]
```diff
@@ -140,30 +143,115 @@ bool emberAfDoorLockClusterLockDoorCallback(chip::app::CommandHandler * commandO
...
+        success = emberAfPluginDoorLockOnDoorLockCommand(endpoint, cmd_pin);
+    }
+
+    emberAfSendImmediateDefaultResponse(success ? EMBER_ZCL_STATUS_SUCCESS : EMBER_ZCL_STATUS_FAILURE);
+
+    // Should be called after reponse sent (indicating command complete) since
+    // the attribute Set call in SetLockState will error out if the command is
+    // still pending
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:186]: Wait, why will the attribute set call error out?

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 12]
```diff
@@ -140,30 +143,115 @@ bool emberAfDoorLockClusterLockDoorCallback(chip::app::CommandHandler * commandO
...
 
-    // TODO: Implement door unlocking by calling emberAfPluginDoorLockOnDoorUnlockCommand
-    emberAfSendImmediateDefaultResponse(EMBER_ZCL_STATUS_SUCCESS);
-    return true;
+    chip::EndpointId endpoint = commandPath.mEndpointId;
+
+    bool require_pin = false;
+    Attributes::RequirePINforRemoteOperation::Get(endpoint, &require_pin);
+
+    if(require_pin)
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:204]: Similar comments throught this function as for LockDoor.

File: src/app/clusters/door-lock-server/door-lock-server.h [hunk 13]
```diff
@@ -34,12 +34,17 @@
 
 class DoorLockServer
 {
+public:
+    // Where should this actually live?
+    char mPin[5];
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.h:39]: In the credential database provided by the app, presumably, with a defined API for the lock to talk to it...

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: .github/workflows/examples-linux-standalone.yaml (hunk 1 above)
Comment (dustin-crossman) [reply .github/workflows/examples-linux-standalone.yaml]: Done

File: examples/lock-app/linux/include/LockManager.h (hunk 2 above)
Comment (dustin-crossman) [reply examples/lock-app/linux/include/LockManager.h]: I was doing what the Linux light app LightingManager class did but I'll change it as you suggest. DoorLockServer uses the Instance() pattern anyway.

File: examples/lock-app/linux/main.cpp (hunk 3 above)
Comment (dustin-crossman) [reply examples/lock-app/linux/main.cpp]: Fixed to just take the bytespan that comes from the command struct.

File: examples/lock-app/linux/main.cpp (hunk 4 above)
Comment (dustin-crossman) [reply examples/lock-app/linux/main.cpp:53]: Looking around at other apps all the *ClusterInitCallback() functions are defined in the app itself so I think this should be here. There is a MatterDoorLockPluginServerInitCallback() that is defined in door-lock-server.cpp that gets called on init.

File: src/app/clusters/door-lock-server/door-lock-server.cpp (hunk 5 above)
Comment (dustin-crossman) [reply src/app/clusters/door-lock-server/door-lock-server.cpp]: Fixed

File: src/app/clusters/door-lock-server/door-lock-server.cpp (hunk 6 above)
Comment (dustin-crossman) [reply src/app/clusters/door-lock-server/door-lock-server.cpp]: Fixed

File: src/app/clusters/door-lock-server/door-lock-server.cpp (hunk 7 above)
Comment (dustin-crossman) [reply src/app/clusters/door-lock-server/door-lock-server.cpp]: Fixed

File: src/app/clusters/door-lock-server/door-lock-server.cpp (hunk 8 above)
Comment (dustin-crossman) [reply src/app/clusters/door-lock-server/door-lock-server.cpp]: Fixed

File: src/app/clusters/door-lock-server/door-lock-server.cpp (hunk 9 above)
Comment (dustin-crossman) [reply src/app/clusters/door-lock-server/door-lock-server.cpp]: Done

File: src/app/clusters/door-lock-server/door-lock-server.cpp (hunk 10 above)
Comment (dustin-crossman) [reply src/app/clusters/door-lock-server/door-lock-server.cpp]: Yea good catch. Reworked the logic.

File: src/app/clusters/door-lock-server/door-lock-server.cpp (hunk 11 above)
Comment (dustin-crossman) [reply src/app/clusters/door-lock-server/door-lock-server.cpp:183]: I was getting the following error from the app output when calling Set (Note: Lines numbers may not match anymore):
[1639081594.128290][29991:29991] CHIP:-: ../../examples/lock-app/linux/third_party/connectedhomeip/src/app/CommandHandler.cpp:389: CHIP Error 0x00000003: Incorrect state at ../../examples/lock-app/linux/third_party/connectedhomeip/src/app/CommandHandler.cpp:308
[1639081594.128345][29991:29991] CHIP:ZCL: Failed to dispatch command, TLVError=../../examples/lock-app/linux/third_party/connectedhomeip/zzz_generated/app-common/app-common/zap-generated/cluster-objects.cpp:9015: Success

The actual line in CommandHandler that was throwing the error was:
// We must not be in the middle of preparing a command, or having prepared or sent one.
VerifyOrReturnError(mState == State::Idle, CHIP_ERROR_INCORRECT_STATE);

The error went away when I moved the Set call after the emberAfSendImmediateDefaultResponse() call.

File: src/app/clusters/door-lock-server/door-lock-server.cpp (hunk 12 above)
Comment (dustin-crossman) [reply src/app/clusters/door-lock-server/door-lock-server.cpp]: Fixed

File: src/app/clusters/door-lock-server/door-lock-server.h (hunk 13 above)
Comment (dustin-crossman) [reply src/app/clusters/door-lock-server/door-lock-server.h]: Made this comment a TODO saying this should be implemented when SetCredential is implemented.

**Comment by dustin-crossman** [comment]
@andy31415 Is there an alternative to running restyle? Running ./scripts/helpers/restyle-diff.sh isn't working for me (some connection error with Docker. Not sure what's wrong).

File: examples/lock-app/linux/main.cpp (hunk 4 above)
Comment (bzbarsky-apple) [reply examples/lock-app/linux/main.cpp:54]: There are two different callbacks.  One is for stack startup init, one is for per-endpoint init.

That said, I just realized this is `emberAfDoorLockClusterInitCallback` which is indeed an app callback, unlike `emberAfDoorLockClusterServerInitCallback` which is the cluster-internal one.  We need better naming....

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 14]
```diff
@@ -12,8 +12,7 @@
  *    distributed under the License is distributed on an "AS IS" BASIS,
  *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  *    See the License for the specific language governing permissions and
- *    limitations under the License.
- */
+ *    limitations under the License. */
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:15]: Why this change?

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 15]
```diff
@@ -140,30 +145,107 @@ bool emberAfDoorLockClusterLockDoorCallback(chip::app::CommandHandler * commandO
...
+    }
+
+    emberAfSendImmediateDefaultResponse(success ? EMBER_ZCL_STATUS_SUCCESS : EMBER_ZCL_STATUS_FAILURE);
+
+    // Should be called after reponse sent (indicating command complete) since
+    // the attribute Set call in SetLockState will error out if the command is
+    // still pending
+    if (success)
+        return DoorLockServer::Instance().SetLockState(endpoint, DlLockState::kLocked) == EMBER_ZCL_STATUS_SUCCESS;
+    return success;
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:187]: Wait, this is not right.  If success is false, we already sent a response above, but returning false here will send one _again_.

A false return only happens when we could not process things enough to even send a failure response, basically.  This should be "return true".  And the SetLockState() call should probably not be returning, if it's after the default response is sent.

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 16]
```diff
@@ -140,30 +145,107 @@ bool emberAfDoorLockClusterLockDoorCallback(chip::app::CommandHandler * commandO
...
+    }
+
+    emberAfSendImmediateDefaultResponse(success ? EMBER_ZCL_STATUS_SUCCESS : EMBER_ZCL_STATUS_FAILURE);
+
+    // Should be called after reponse sent (indicating command complete) since
+    // the attribute Set call in SetLockState will error out if the command is
+    // still pending
+    if (success)
+        return DoorLockServer::Instance().SetLockState(endpoint, DlLockState::kUnlocked) == EMBER_ZCL_STATUS_SUCCESS;
+    return success;
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:233]: Similar issue here.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/clusters/door-lock-server/door-lock-server.cpp (hunk 11 above)
Comment (bzbarsky-apple) [reply src/app/clusters/door-lock-server/door-lock-server.cpp:184]: I just tried putting the Set before the send, and it was working fine.  Please describe the exact steps to reproduce that show this problem?  Exact commands executed in which exact client, etc.

Comment (dustin-crossman) [reply src/app/clusters/door-lock-server/door-lock-server.cpp:184]: Well I can't seem to reproduce the issue. I tried building like you did (for onnetwork pairing) as well as using two Pis and running ble-wifi pairing. I was working off an older master when I saw the issue so perhaps it's been fixed since.

File: src/app/clusters/door-lock-server/door-lock-server.cpp (hunk 14 above)
Comment (dustin-crossman) [reply src/app/clusters/door-lock-server/door-lock-server.cpp:15]: Not sure. I was thinking it might have been clang-format but it doesn't seem touch that line either way. I'll change it back.

File: src/app/clusters/door-lock-server/door-lock-server.cpp (hunk 11 above)
Comment (bzbarsky-apple) [reply src/app/clusters/door-lock-server/door-lock-server.cpp]: In that case, it's probably better to update the internal state before sending the response.

File: src/app/clusters/door-lock-server/door-lock-server.cpp (hunk 15 above)
Comment (dustin-crossman) [reply src/app/clusters/door-lock-server/door-lock-server.cpp]: Ok. Reworked to send default response after SetLockState is called (and return checked) and just have the callback return true.

File: src/app/clusters/door-lock-server/door-lock-server.cpp (hunk 16 above)
Comment (dustin-crossman) [reply src/app/clusters/door-lock-server/door-lock-server.cpp]: Same fix as above.

**Review by bzbarsky-apple — APPROVED** [review]

**Comment by bzbarsky-apple** [comment]
@dustin-crossman Need to pick up restyle changes?

**Comment by dustin-crossman** [comment]
@bzbarsky-apple Restyled!

**Comment by bzbarsky-apple** [comment]
@dustin-crossman the QPG failure is real.  Needs to compile in some .cpp file with that function?

**Comment by dustin-crossman** [comment]
Added some weak definitions for those door lock/unlock functions.

File: .github/workflows/examples-linux-arm.yaml [hunk 17]
```diff
@@ -67,6 +67,7 @@ jobs:
                         --target linux-arm64-chip-tool-ipv6only \
                         --target linux-arm64-minmdns \
                         --target linux-arm64-thermostat-no-ble \
+                        --target linux-arm64-lock \
```
Comment (andy31415) [inline .github/workflows/examples-linux-arm.yaml:70]: do we need to validate the arm cross compile? Trying to minimize the amount of CI time - we do not build light for example, just want to have a subset here.

Unless there is reason to believe the arm64 build will show things that the x64 will not (and nor will things like all-clusters) please remove. The other builds try to mix variants (all clusters is everything, ipv6only and no-ble variants are split, minmdns validates dns clients and servers)

File: examples/lock-app/linux/include/LockManager.h [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+class LockManager
+{
+public:
+    void Init();
+
+    bool Lock(chip::Optional<chip::ByteSpan> pin);
+    bool Unlock(chip::Optional<chip::ByteSpan> pin);
+
+    static LockManager & Instance();
+    static LockManager instance;
```
Comment (andy31415) [inline examples/lock-app/linux/include/LockManager.h:37]: instance should not be public - the `Instance()` call is.

File: examples/lock-app/linux/src/LockManager.cpp [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+LockManager LockManager::instance;
+
+LockManager & LockManager::Instance()
+{
+    return instance;
+}
+
+void LockManager::Init()
+{
+    mLocked = false;
```
Comment (andy31415) [inline examples/lock-app/linux/src/LockManager.cpp:34]: set mLocked in the constructor (i.e. in the header).

Do we even need an Init at this point?

File: .github/workflows/examples-linux-arm.yaml (hunk 17 above)
Comment (dustin-crossman) [reply .github/workflows/examples-linux-arm.yaml]: Removed!

File: examples/lock-app/linux/include/LockManager.h (hunk 18 above)
Comment (dustin-crossman) [reply examples/lock-app/linux/include/LockManager.h]: Done

File: examples/lock-app/linux/src/LockManager.cpp (hunk 19 above)
Comment (dustin-crossman) [reply examples/lock-app/linux/src/LockManager.cpp]: Ok. Init removed.

**Comment by dustin-crossman** [comment]
@andy31415 @bzbarsky-apple Had a number of build errors but they kind of looked like flaky CI so I rebased to master again. Hopefully that clears the build errors up.

**Comment by bzbarsky-apple** [comment]
The K32W failures look like real errors in code: wrong format strings being used.  Should be using PR* formats for *int*_t types.

The "Build on Linux" failure seems like a real test failure too; presumably test expectations for the build script need updating?

The mbed failure is because `MatterDoorLockPluginServerInitCallback` is not defined.  Presumably on that platform the door lock cluster .cpp needs to be included in the build?  This is likely to be true for all the non-linux door lock apps, since they all share the same .zap file so have the cluster enabled now.

I didn't look at the other failures.  But those three are not flakes.

**Review by jrhees-cae — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]
